contrib/win32/win32compat/fileio.c

```diff
@@ -749,13 +752,98 @@ fileio_fstat(struct w32_io* pio, struct _stat64 *buf)
 	return _fstat64(fd, buf);
 }
 
+wchar_t * fileio_readlink_internal(wchar_t * wpath)
+{
+	/* abbreviated reparse data structure for symlinks */
+	typedef struct _REPARSE_DATA_BUFFER_SYMLINK {
+		ULONG ReparseTag;
+		USHORT ReparseDataLength;
+		USHORT Reserved;
+		USHORT SubstituteNameOffset;
+		USHORT SubstituteNameLength;
+		USHORT PrintNameOffset;
+		USHORT PrintNameLength;
+		ULONG Flags;
+	} REPARSE_DATA_BUFFER, *PREPARSE_DATA_BUFFER_SYMLINK;
+
+	/* early declarations for cleanup */
+	wchar_t* linkpath = NULL;
+	HANDLE handle = INVALID_HANDLE_VALUE;
+	PREPARSE_DATA_BUFFER_SYMLINK reparse_buffer = NULL;
+
+	/* obtain a handle to send to deviceioctl  */
+	handle = CreateFileW(wpath, 0, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+		0, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT, 0);
+	if (handle == INVALID_HANDLE_VALUE) {
+		errno = errno_from_Win32LastError();
+		goto cleanup;
+	}
+
+	/* send a request to the file system to get the real path */
+	reparse_buffer = (PREPARSE_DATA_BUFFER_SYMLINK) malloc(MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+	DWORD dwBytesReturned = 0;
+	if (DeviceIoControl(handle, FSCTL_GET_REPARSE_POINT, NULL, 0,
+		(LPVOID)reparse_buffer, MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &dwBytesReturned, 0) == 0) {
+		errno = errno_from_Win32LastError();
+		goto cleanup;
+	}
+
+	/* ensure file is actually symlink */
+	if (reparse_buffer->ReparseTag != IO_REPARSE_TAG_SYMLINK) {
+		errno = ENOLINK;
+		goto cleanup;
+	}
+
+	/* trim off the extended drive prefix if present */
+	const wchar_t prefix[] = L"\\??\\";
```

Comment:
See this: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx.  I'll better characterize this in my code.

Reply:
Have you tried a symbolic link referencing a shared path? If yes, does it start with prefix "\\?\".

In code you used "\\?\\" instead of "\\?\", is it intentional? If yes, why?

Reply:
Good catch.  It comes back as \??\UNC\ instead of \??\C:\.  I will have to address this as well.

Reply:
Addressed in latest commit.